log/logging.go

```diff
@@ -188,22 +214,123 @@ func SetTimestampFlags(flags int) {
+	Error(description string, keysAndValues ...interface{})
+	Warn(description string, keysAndValues ...interface{})
+	Info(description string, keysAndValues ...interface{})
+	Debug(description string, keysAndValues ...interface{})
+	Trace(description string, keysAndValues ...interface{})
+	SetOutput(w io.Writer)
+	SetTimestampFlags(flags int)
+	SetField(name string, value interface{})
+}
+
+func NewLogger(format LogFormat, id string, staticKeysAndValues ...interface{}) LoggerInterface {
+	return newLoggerStruct(format, id, staticKeysAndValues...)
+}
+
+func newLoggerStruct(format LogFormat, id string, staticKeysAndValues ...interface{}) *Logger {
+	var prefix string
+	var flags int
+	staticArgs := make(map[string]string, 0)
+
+	format = SanitizeFormat(format)
+
+	if format == JsonFormat {
+		// Don't mess up the json by letting logger print these:
+		prefix = ""
+		flags = 0
+
+		// Instead put them into the staticArgs
+		if defaultPrefix != "" {
+			// Try to interpret at json, to get structured data.
+			var prefixData map[string]interface{}
+			if err := json.Unmarshal([]byte(defaultPrefix), &prefixData); err == nil {
+				for key, value := range prefixData {
+					staticArgs[key] = fmt.Sprintf("%v", value)
+				}
+			} else {
+				// Oh well, just use it as a single item.
+				staticArgs["prefix"] = defaultPrefix
+			}
+		}
+	} else {
+		prefix = defaultPrefix
+		flags = Flags
+	}
+
+	// Do this after handling json prefix, cuz individual loggers should override
+	// external env variable.
+	currentKey := ""
+	for i, arg := range staticKeysAndValues {
+		if i%2 == 0 {
+			currentKey = fmt.Sprintf("%v", arg)
+		} else {
+			staticArgs[currentKey] = fmt.Sprintf("%v", arg)
+		}
+	}
+
+	// If there are an odd number of keys+values, add the dangling key with empty
+	// value.
+	if len(staticKeysAndValues)%2 == 1 {
+		staticArgs[currentKey] = ""
+	}
+
+	return &Logger{
+		ID:    id,
+		Level: Level,
+
+		format:     format,
+		staticArgs: staticArgs,
+
+		// don't touch the default logger on 'log' package
+		l: log.New(defaultOutput, prefix, flags),
+	}
+}
+
+type LogFormat string
+
+const (
+	DefaultFormat   LogFormat = "" // Use env variable, defaulting to PlainTextFormat
+	PlainTextFormat           = "text"
+	JsonFormat                = "json"
+)
+
+func SanitizeFormat(format LogFormat) LogFormat {
+	if format == PlainTextFormat || format == JsonFormat {
+		return format
+	} else {
+		// Whether it's explicitly a DefaultFormat, or it's an unrecognized value,
+		// try to take from env var.
+		envFormat := os.Getenv("DEFAULT_LOG_ENCODING_FORMAT")
+		if envFormat == string(JsonFormat) || envFormat == string(PlainTextFormat) {
+			return LogFormat(envFormat)
+		}
+	}
+
+	// Fall back to text
+	return PlainTextFormat
+}
+
 // New creates a new logger instance.
+// DEPRECATED: use `NewLogger(...)` instead. That one returns an interface,
+// which allows the underlying data structure to change without breaking
+// clients.
 func New() *Logger {
```

Comment:
I'd take a more radical approach and make `Logger` an interface and `New()` return that interface. The current struct would then become `logger`.

Also, by moving to the interface approach, I'd use the format to create a different type of logger per format, instead of having a single struct handle all possible types of formats, i.e. `PlainTextFormat` -> `plainTextLogger` struct, `JsonFormat` -> `jsonLogger` struct.